azurerm/internal/services/dataprotection/data_protection_backup_instance_postgresql_resource.go

```diff
@@ -0,0 +1,223 @@
+	defer cancel()
+
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+	vaultName := d.Get("vault_name").(string)
+
+	id := parse.NewBackupInstanceID(subscriptionId, resourceGroup, vaultName, name)
+
+	if d.IsNewResource() {
+		existing, err := client.Get(ctx, id.BackupVaultName, id.ResourceGroup, id.Name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for existing DataProtection BackupInstance (%q): %+v", id, err)
+			}
+		}
+		if !utils.ResponseWasNotFound(existing.Response) {
+			return tf.ImportAsExistsError("azurerm_data_protection_backup_instance_postgresql", id.ID())
+		}
+	}
+
+	databaseId, _ := postgresParse.DatabaseID(d.Get("database_id").(string))
+	location := location.Normalize(d.Get("location").(string))
+	serverId := postgresParse.NewServerID(databaseId.SubscriptionId, databaseId.ResourceGroup, databaseId.ServerName)
+	policyId, _ := parse.BackupPolicyID(d.Get("backup_policy_id").(string))
+
+	parameters := dataprotection.BackupInstanceResource{
+		Properties: &dataprotection.BackupInstance{
+			DataSourceInfo: &dataprotection.Datasource{
+				DatasourceType:   utils.String("Microsoft.DBforPostgreSQL/servers/databases"),
+				ObjectType:       utils.String("Datasource"),
+				ResourceID:       utils.String(databaseId.ID()),
+				ResourceLocation: utils.String(location),
+				ResourceName:     utils.String(databaseId.Name),
+				ResourceType:     utils.String("Microsoft.DBforPostgreSQL/servers/databases"),
+				ResourceURI:      utils.String(""),
+			},
+			DataSourceSetInfo: &dataprotection.DatasourceSet{
+				DatasourceType:   utils.String("Microsoft.DBForPostgreSQL/servers"),
+				ObjectType:       utils.String("DatasourceSet"),
+				ResourceID:       utils.String(serverId.ID()),
+				ResourceLocation: utils.String(location),
+				ResourceName:     utils.String(serverId.Name),
+				ResourceType:     utils.String("Microsoft.DBForPostgreSQL/servers"),
+				ResourceURI:      utils.String(""),
+			},
+			FriendlyName: utils.String(id.Name),
+			PolicyInfo: &dataprotection.PolicyInfo{
+				PolicyID: utils.String(policyId.ID()),
+			},
+		},
+	}
+	future, err := client.CreateOrUpdate(ctx, id.BackupVaultName, id.ResourceGroup, id.Name, parameters)
+	if err != nil {
+		return fmt.Errorf("creating/updating DataProtection BackupInstance (%q): %+v", id, err)
+	}
+
+	if err := future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("waiting for creation/update of the DataProtection BackupInstance (%q): %+v", id, err)
+	}
+
+	stateConf := &pluginsdk.StateChangeConf{
+		Pending:    []string{string(dataprotection.ConfiguringProtection), string(dataprotection.UpdatingProtection)},
+		Target:     []string{string(dataprotection.ProtectionConfigured)},
+		Refresh:    policyProtectionStateRefreshFunc(ctx, client, id),
+		MinTimeout: 1 * time.Minute,
+		Timeout:    d.Timeout(pluginsdk.TimeoutCreate),
+	}
```

Comment:
this should be using the remaining time from the context, since these timeouts will start at different times otherwise

```suggestion
	deadline, ok := ctx.Deadline()
	if !ok {
		return fmt.Errorf("context had no deadline")
	}
	stateConf := &pluginsdk.StateChangeConf{
		Pending:    []string{string(dataprotection.ConfiguringProtection), string(dataprotection.UpdatingProtection)},
		Target:     []string{string(dataprotection.ProtectionConfigured)},
		Refresh:    policyProtectionStateRefreshFunc(ctx, client, id),
		MinTimeout: 1 * time.Minute,
		Timeout:    time.Until(deadline),
	}
```